フォロー機能を実装(DAIKI)

i-k-i-a-d opened the pull request:
・多対多のDBを作るためにマイグレーションファイルを作ってfollow_usersテーブルを中間テーブルとして作成。
・FollowUserモデルを作成。
・Userモデル内に多対多のリレーションメソッドを作成。
・フォローとフォロー解除機能のためのルーティング作成。
・ルーティング通りにFollowUserControllerを新たに作成してメソッドを二つ作成。
・フォロー・フォロー解除のボタンをユーザー詳細画面に設置して、条件分岐で表示を切り替えるように実装。
・ユーザーが退会した際にそのユーザーがフォローしたりフォローされたりしているデータも中間テーブルから削除されるように実装。

ユーザーが退会した際は、そのユーザーがフォローしていたデータとフォローされていたデータが全て中間テーブルから物理削除される仕様にしておりますがよろしいでしょうか？？

今回のタスクは、お手本のアプリにも機能がなく、全て手探りで進めた実装になりますので、もっと効率的な実装方法があるかもしれませんが、とりあえず要件通りに動くようには仕上げております。

レビューよろしくお願いいたします。

i-k-i-a-d committed:
多対多のDBを作るためにマイグレーションファイルを作ってfollow_usersテーブルを中間テーブルとして作成

FollowUserモデルを作成。
Userモデル内に多対多のリレーションメソッドを作成。
フォローとフォロー解除機能のためのルーティング作成。
ルーティング通りにFollowUserControllerを新たに作成してメソッドを二つ作成。
フォロー・フォロー解除のボタンをユーザー詳細画面に設置して、条件分岐で表示を切り替えるように実装。
ユーザーが退会した際にそのユーザーがフォローしたりフォローされたりしているデータも中間テーブルから削除されるように実装。

Osk0830 commented on app/FollowUser.php:
```diff
@@ -0,0 +1,13 @@
+<?php
+
+namespace App;
+
+use Illuminate\Database\Eloquent\Model;
+use Illuminate\Database\Eloquent\Relations\Pivot;
+
+class FollowUser extends Pivot
+{
+    protected $fillable = ['following_user_id', 'followed_user_id'];
+
+    protected $table = 'follow_users';
+}
```
- ちょっと抽象的な話になってしまいますが、モデルは、ユーザーや投稿のように「物体」が存在する際に、その物体の雛形として存在するイメージです。中間テーブルはあくまで2つのモデルの「関係」を整理するためだけのものなので、モデルは不要です。

i-k-i-a-d replied:
FollowUserモデルを削除すると、ユーザー退会時のフォローの削除が行えなくなるため消さずに残しております。
また、attach/detachを使って中間テーブルに値を入れるようにしたので、$fillableの設定は不要となり、削除しております。

Osk0830 commented on app/Http/Controllers/FollowUserController.php:
```diff
@@ -0,0 +1,46 @@
+     * @param string $id
+     * @return view
+     */
+    public function follow($id) {
+        DB::BeginTransaction();
+        try {
+            FollowUser::create([
+                'following_user_id' => Auth::id(),
+                'followed_user_id' => $id,
+            ]);
+            DB::commit();
+        } catch (\Throwable $th) {
+            DB::rollBack();
+            abort(500);
+        }
+        return redirect(route('users.show', $id));
+    }
+
+    /**
+     * フォローを解除する処理。
+     * @param string $id
+     * @return view
+     */
+    public function unfollow($id) {
+        try {
+            $follow = FollowUser::where('following_user_id', '=', Auth::id())->where('followed_user_id', '=', $id)->first();
+            $follow->delete();
+            return redirect(route('users.show', $id));
+        } catch (\Throwable $th) {
+            abort(500);
+        }
+    }
```
- 中間テーブルにデータを入れたり中間テーブルのデータを削除するのには、`attach／detach`というメソッドを用いるのが一般的です。リレーションメソッドとうまく組み合わせながら、`attach／detach`を活用して実装してみてください。
- なお、上記のような操作は、他の場所でも使いまわせるよう、モデルのメソッドとして定義し、コントローラーの各アクション内ではそれを呼び出すような形で記述するのがいいでしょう。ユーザーモデル内にメソッドを定義し、ここではそれを呼び出す形にしてみてくださいね。

i-k-i-a-d replied:
attach/detachを使って修正し、場所をUserモデル内に引っ越しました。
また、FollowUserControllerから静的なメソッドとして呼び出すように実装してます。

Osk0830 commented on app/Http/Controllers/UserController.php:
```diff
@@ -60,7 +61,8 @@ public function updateUser (UserRequest $request) {
     public function showDetail ($id) {
         $user = User::find($id);
-        return view('users.detail', ['user' => $user]);
+        $follow = FollowUser::where('following_user_id', "=", Auth::id())->where('followed_user_id', "=", $id)->first();
```
- このような情報はリレーションメソッドを活用して取得するようにしましょう。ただ、コントローラーで取得し渡す必要はなく、Blade内で取得する感じで良さそうですね。

i-k-i-a-d replied:
リレーションメソッドを使って修正しました。

Osk0830 commented on app/Http/Controllers/UserController.php:
```diff
@@ -72,9 +74,17 @@ public function deleteUser ($id) {
         if (Auth::id() === (int)$id) {
             try {
                 User::find($id)->delete();
+                $follows = FollowUser::where('following_user_id', '=', $id)->get();
+                foreach ($follows as $follow) {
+                    FollowUser::find($follow['id'])->delete();
+                }
+                $follows = FollowUser::where('followed_user_id', '=', Auth::id())->get();
+                foreach ($follows as $follow) {
+                    FollowUser::find($follow['id'])->delete();
+                }
```
- このような処理は、マイグレーションにある `onDelete('cascade')`により自動的に行われるようになるので、記述が不要ですね。

i-k-i-a-d replied:
usersテーブルが論理削除の場合はマイグレーションにある onDelete('cascade')だけでは削除できないので、そのままの状態にしております。

Osk0830 commented on resources/views/users/detail.blade.php:
```diff
@@ -8,10 +8,32 @@
-                    <div class="mt-3">
-                        <a href="{{ route('users.edit', $user->id) }}" class="btn btn-primary btn-block">ユーザ情報の編集</a>
-                    </div>
+                @if(Auth::check())
+                    @if (!is_null($follow))
+                        @if (Auth::id() === $user->id)
+                            <div class="mt-3">
+                                <a href="{{ route('users.edit', $user->id) }}" class="btn btn-primary btn-block">ユーザ情報の編集</a>
+                            </div>
+                        @else
+                            <div class="mt-3">
+                                <form method="POST" action="{{ route('unfollow', $user->id) }}">
+                                    @csrf
+                                    <button type="submit" class="btn btn-primary btn-block">フォローを解除する</button>
+                                </form>
+                            </div>
+                        @endif
+                    @else 
+                        @if (Auth::id() === $user->id)
+                            <div class="mt-3">
+                                <a href="{{ route('users.edit', $user->id) }}" class="btn btn-primary btn-block">ユーザ情報の編集</a>
+                            </div>
+                        @else
+                            <form method="POST" action="{{ route('follow', $user->id) }}">
+                                @csrf
+                                <button type="submit" class="btn btn-primary btn-block">フォローする</button>
+                            </form>
+                        @endif
+                    @endif 
                 @endif
```
- モデル・コントローラーを変更・調整した結果に即して必要箇所調整くださいね。

i-k-i-a-d replied:
条件分岐の内容をリレーションメソッドを使って書き換えました。

Osk0830 commented on routes/web.php:
```diff
@@ -22,6 +22,8 @@
 Route::get('users/{id}/edit', 'UserController@showEdit')->name('users.edit');
 Route::post('update/user', 'UserController@updateUser')->name('users.update');
 Route::post('users/delete/{id}', 'UserController@deleteUser')->name('users.delete');
+Route::post('users/{id}/follow', 'FollowUserController@follow')->name('follow');
+Route::post('users/{id}/unfollow', 'FollowUserController@unfollow')->name('unfollow');
```
- この機会にルーティングの整理も行ってみてください。その際、必要な箇所にはミドルウェアも挟みましょう。

i-k-i-a-d replied:
prefixを使って修正しました。

Osk0830 reviewed (changes requested):
お疲れ様でした！
色々と考えながら実装を進められているのが伺えて非常に素晴らしいと思います。
ただ、フレームワークを用いる実装で重要になってくるポイントの一つは、フレームワークが用意している機能を活用していかにスマートにスリムに書けるか、という部分にもあります。
レビュー内容を踏まえて一度検討、修正してみてくださいね。

i-k-i-a-d committed:
フォロー・フォロー解除のロジックをattach/detachを使った実装に修正。
それに伴い、FollowUserモデル内のが不要となり削除。
フォロー・フォロー解除のロジックをUserモデルへ引っ越し。
FollowUserControllerから静的なメソッドとしてフォロー・フォロー解除のメソッドを呼び出せるよう実装。
ユーザー詳細画面のフォロー・フォロー解除ボタンの表示の条件分岐をリレーションメソッドを使って修正。
ルーティングをprefixを使って整理。

i-k-i-a-d committed:
新規ユーザー登録の実装をdevelopにpullしてfeatureブランチにmergeしてコンフリクト解消。

Osk0830 commented on app/User.php:
```diff
@@ -52,4 +55,47 @@ public static function boot()
+     * @param string $id
+     * @return view
+     */
+    public static function follow($id) {
+        DB::BeginTransaction();
+        try {
+            $user = User::find($id);
+            $user->followUsers()->attach([
+                'following_user_id' => Auth::id(),
+            ]);
+            DB::commit();
+        } catch (\Throwable $e) {
+            DB::rollBack();
+            abort(500);
+        }
+    }
+
+    /**
+     * フォローを解除する処理。
+     * @param string $id
+     * @return view
+     */
+    public static function unfollow($id) {
+        try {
+            $user = User::find($id);
+            $user->followUsers()->detach();
+        } catch (\Throwable $e) {
+            abort(500);
+        }
+    }
```
なかなかいい感じになってきましたね。ただ、いくつか再検討や修正をした方が良さそうな内容があります。

- フォローという操作的に考えた場合、フォローボタンを押下したユーザーを軸に、自分のidにフォロー相手のidを紐づける、という論理が順当でわかりやすいと思いますが、今はわざわざフォロー相手の情報を取得して、そのユーザーを軸に、ログインしているユーザーのidを紐づける、というしくみになってしまっています。ここら辺の論理と実装をもう少しスマートにしてみてください。

- 現状では、下記のように自分自身であった場合もフォローできてしまいます。自分自身はフォローできない仕組みにしましょう。

[![Image from Gyazo](https://i.gyazo.com/0f97cbcdfea7f7b7ed9b43da5d7e53a0.png)](https://gyazo.com/0f97cbcdfea7f7b7ed9b43da5d7e53a0)

[![Image from Gyazo](https://i.gyazo.com/19d79fe7ca5e637caae3f34aed3d6074.png)](https://gyazo.com/19d79fe7ca5e637caae3f34aed3d6074)

- また、フォロー関係（特定のidのペア）がすでに存在している場合にもフォローできてしまいます。同じ理屈で、関係が存在しなくても `detach()`が実行される仕組みになっています。すでにフォローしているかどうか（テーブルにidのペアが存在するか）に応じて処理が分岐されるようにしてみてください。

[![Image from Gyazo](https://i.gyazo.com/0fe6d4862ccbd350eea3ede1768c89c4.png)](https://gyazo.com/0fe6d4862ccbd350eea3ede1768c89c4)

- 既にフォローしているかどうかをチェックするロジックは、モデル内にメソッドとして定義しておくと、Bladeなどでもそれを利用でき記述がシンプルになると思います。

i-k-i-a-d replied:
マイグレーションファイルでfollow_usersテーブル内のカラム名を修正。
特定のユーザーを基準にして、そのユーザーがフォローしているユーザー情報を全て取得するリレーションメソッドと、特定のユーザーがフォローされているユーザー情報を全て取得するリレーションメソッドを作成。
そのリレーションメソッドを使って、ログインしているユーザーを基準として、フォローするとログインしているユーザーのidとフォローされたユーザーのidがテーブルに挿入されるように実装。
また、FollowUserController内でログインユーザーとフォローされるユーザーのidが異なるときしかメソッドが実行されないように条件分岐を追加。

Osk0830 commented on database/migrations/2023_08_17_174711_create_follow_users_table.php:
```diff
@@ -0,0 +1,35 @@
+     */
+    public function up()
+    {
+        Schema::create('follow_users', function (Blueprint $table) {
+            $table->bigIncrements('id');
+            $table->unsignedBigInteger('followed_user_id')->index();
+            $table->unsignedBigInteger('following_user_id')->index();
+            $table->foreign('followed_user_id')->references('id')->on('users')->onDelete('cascade');
+            $table->foreign('following_user_id')->references('id')->on('users')->onDelete('cascade');
+            $table->timestamps();
```
- DBにも、idのペアはユニークなもので重複不可のように制約をかけておきましょう。

i-k-i-a-d replied:
idのペアがユニークになるようにマイグレーションファイルで設定。

Osk0830 commented on resources/views/users/detail.blade.php:
```diff
@@ -8,10 +8,24 @@
+                            <div class="mt-3">
+                                <a href="{{ route('users.edit', $user->id) }}" class="btn btn-primary btn-block">ユーザ情報の編集</a>
+                            </div>
+                    @elseif ($user->followUsers()->where('following_user_id', Auth::id())->exists() && $user->followUsers()->where('followed_user_id', $user->id)->exists())
```
- 上述したように、モデルに必要なメソッドを用意してあげることで、ここら辺の記述ももう少しシンプルに書けると思います。

i-k-i-a-d replied:
Userモデル内にフォロー状態をチェックできるメソッドを作って、$userインスタンスからbladeファイル内で呼び出して条件分岐できるように実装。

Osk0830 commented on routes/web.php:
```diff
@@ -15,16 +15,29 @@
+    Route::post('/', 'Auth\LoginController@login')->name('login');
+});
+
+Route::prefix('users')->group(function () {
+    Route::prefix('{id}')->group(function () {
+        Route::get('/', 'UserController@showDetail')->name('users.show'); 
+        Route::get('edit', 'UserController@showEdit')->name('users.edit'); 
+        Route::post('delete', 'UserController@deleteUser')->name('users.delete'); 
+        Route::post('follow', 'FollowUserController@exeFollow')->name('follow'); 
+        Route::post('unfollow', 'FollowUserController@exeUnfollow')->name('unfollow');
+    });
+    Route::post('update', 'UserController@updateUser')->name('users.update');
+});
```
- ログインしている場合しかアクセスできないようにする必要がある内容には、ミドルウェアを挟んでおきましょう。
- `update`って、id無関係でしたっけ？？

i-k-i-a-d replied:
ミドルウェアでログイン状態のルーティングをまとめました。

i-k-i-a-d replied:
updateに関しては、ルートパラメーターでidを受け取らなくても、inputタグから直接Requestクラスへ、更新するユーザーのidが届くようになってます。

Osk0830 reviewed (changes requested):
お疲れ様でした。
記述が比較的スマートになってきましたね。

モデルの問題と削除の問題は一旦保留し、それ以外について幾つかコメントしました。
確認よろしくお願いします。

i-k-i-a-d committed:
マイグレーションファイルで中間テーブルのカラム名を修正。
マイグレーションファイルに複合ユニークを設定。
Userモデル内のフォロー・フォロー解除メソッドも修正。
ルーティングにミドルウェアを実装してログイン時しか機能しないルーティングを実装。
Userモデル内のフォロー機能用の多対多リレーションメソッドを新しいカラム名で修正。
FollowUserController内のフォロー・フォロー解除メソッドを呼び出すメソッドにフォローする側とフォローされる側が一緒ではないときしか実行されないように条件分岐実装。
加えて、フォローしていないときしか、フォローされないように実装。その逆も実装。
Userモデル内にフォローチェック用のメソッドを作成。
ユーザー退会時のフォローデータの削除をFollowUserモデルを使わずにリレーションメソッドとフォロー解除メソッドで実装。
FollowUserモデルがいよいよ不要になったのでFollowUser.phpファイルごと削除。

i-k-i-a-d committed:
Merge branch 'develop_b_fumiduki_rab' into feature/daiki/implement_follow

i-k-i-a-d committed:
フォローチェックの条件分岐とフォローする側とされる側が同一ユーザーでないかどうかの条件分岐を直接followメソッドへ引っ越し。
同じ条件分岐をunfollowメソッドへは実装しないことにした。ユーザー退会の際に使いまわせなくなるので。
ルーティングの整理を更に強化した。
動作確認を行う際にDBへ値を入れる工数削減のために中間テーブル用のSeederファイルを作成した。
ユーザーの削除はフォローデータの削除が行われた後に実行されるようにプログラムの順番を変更した。

Osk0830 commented on app/Http/Controllers/FollowUserController.php:
```diff
@@ -0,0 +1,26 @@
+        $user->unfollow($id);
+        return redirect(route('users.show', $id));
+        abort(404);
+    }
```
- `abort(404)`は`return`の後に置かれているので、必ず`return`で処理完了してしまい、ここまで到達することはなくないでしょうか？

i-k-i-a-d replied:
abort関数をfollowメソッド側に引っ越しさせ、条件分岐にはじかれた際の処理として実装してます。

Osk0830 commented on app/User.php:
```diff
@@ -52,4 +55,58 @@ public static function boot()
+     * @param string $id
+     * @return view
+     */
+    public function followCheck ($id) {
+        $user = User::find($id);
+        return $user->followers()->where('user_id', Auth::id())->exists();
+    }
```
- 「フォローボタンを押下した人（ログインしているユーザー）が引数として渡されたidに該当するユーザーを既にフォローしているかどうか」をチェックするものになるべきかな、と思いますが、こちら軸となるユーザーが変わってしまってませんでしょうか？今インスタンス化されているユーザー（ログインユーザー）が軸である場合、`$this`を活用できそうですね。

i-k-i-a-d replied:
ログインユーザーを基準としてリレーションメソッドを使うように修正してます。

Osk0830 commented on app/User.php:
```diff
@@ -52,4 +55,58 @@ public static function boot()
+    public function follow($id) {
+        DB::BeginTransaction();
+        try {
+            $bool = false;
+            if (Auth::id() !== $id && !$bool = $this->followCheck($id)) {
```
- この `$bool`はなぜ必要なのでしょうか？

i-k-i-a-d replied:
必要ありませんでしたね。
$thisの前に！を付けるのが、なんか不自然な気がして変数を置いてみました。

Osk0830 commented on app/User.php:
```diff
@@ -52,4 +55,58 @@ public static function boot()
+        DB::BeginTransaction();
+        try {
+            $bool = false;
+            if (Auth::id() !== $id && !$bool = $this->followCheck($id)) {
+                $this->followings()->attach($id);
+                DB::commit();
+            }
```
- if文の条件に合致しなかった場合の処理はどうなりますか？

i-k-i-a-d replied:
abort関数をfollowメソッドに実装済みです。

Osk0830 commented on app/User.php:
```diff
@@ -52,4 +55,58 @@ public static function boot()
+     * @param string $id
+     * @return view
+     */
+    public function unfollow($id) {
+        try {
+            $this->followings()->detach($id);
+        } catch (\Throwable $e) {
+            abort(500);
+        }
+    }
```
- こちらにも必要なチェックを挟むようにしましょう。

i-k-i-a-d replied:
今回、ユーザー退会時にフォローデータが削除される仕様を実現するのに、このunfollow();メソッドを呼び出して削除させるように実装したので、あえて、ここのunfollow();メソッドには条件分岐を入れませんでした。

理由としては、
そもそも、自分で自分をフォローすることはできないようにfollow();メソッドで条件設定しているため、自分で自分のフォローを解除するという事象自体が起き得ないし、フォロー状態かどうかを確認してフォロー状態でないときは、中間テーブルにデータが存在しないときなので、unfollow();のメソッドが動いてもなにも変化がない。そして最大の理由が、unfollow();メソッドをユーザー退会処理の際にフォローデータを削除する際に使いまわせなくなる。
なぜなら、フォローチェックはログインしているユーザーがフォローしているユーザーの中に、ユーザー詳細画面に表示されているユーザーのidが入っているかどうかを確認する仕様なので、退会する処理の際はログインしているユーザーもユーザー詳細画面に表示しているユーザーも同じユーザーであり、これが成立する場合、自分で自分をフォローしていることになってしまう。
また、そもそもログインしている人が、ユーザー詳細画面に表示されているユーザーではないとき、という条件があるので、完全に条件の範囲外になる。

以上の理由でunfollow();には条件分岐を付けませんでしたが、今回はユーザー退会時のフォロー解除のロジックを別途で構築したのでunfollow();メソッドは使わずに対応しました。

なので、unfollow();メソッドに条件分岐を入れても問題はなくなり、ご指摘通りの実装となっております。

Osk0830 commented on database/migrations/2023_08_17_174711_create_follow_users_table.php:
```diff
@@ -0,0 +1,36 @@
+            $table->bigIncrements('id');
+            $table->unsignedBigInteger('user_id')->index();
+            $table->unsignedBigInteger('followed_id')->index();
+            $table->foreign('user_id')->references('id')->on('users')->onDelete('cascade');
+            $table->foreign('followed_id')->references('id')->on('users')->onDelete('cascade');
+            $table->unique(['user_id', 'followed_id']);
```
- この3つは、カラムそのものの作成ではないので、カラム作成に関するメソッドが全て終わってから1行空けて書くようにすると、より可読性が高まりそうですね。

i-k-i-a-d replied:
カラム生成に関係ないものは一行スペースを空けてまとめました。

Osk0830 commented on routes/web.php:
```diff
@@ -14,17 +14,38 @@
+});
+
+Route::prefix('users')->group(function () {
+    Route::group(['middleware' => 'auth'], function () {
+        Route::post('update', 'UserController@updateUser')->name('users.update');
+    }); 
```
- なるほど、`update`では`hidden`で渡している旨了解です。もちろんこのままでも間違いはないですが、統一性という観点から少しわかりにくいのもあるので、可能であればパラメーターで渡す形に統一するといいかもですね。今ではなくおいおいで全然OKですので、検討してみてくださいね。

i-k-i-a-d replied:
今回ついでなので、やってしまいました。もともとのinputタグの削除、ルーティングの修正、route();関数の第二引数の指定、UserController内のメソッドで二つ目の引数を受取り、それを使って中のロジックの修正まで完了してます。

Osk0830 reviewed (changes requested):
お疲れ様です！
かなりいい感じになってきたと思います！

少し検討が必要な箇所がありコメントしましたので、
確認よろしくお願いします！

i-k-i-a-d committed:
abort関数をfollowメソッド側に引っ越しさせ、条件に合わなかったときの処理を記述。
followメソッド内に既述したabort関数が条件に一致してDBの処理が終わった後に実行されないようにreturn;を追記。
followCheckメソッドのロジックをログインしているユーザーを基準としてリレーションメソッドを使う仕様に変更。
followメソッド内の条件分岐に利用した変数は不要だったので削除。
unfollowメソッドには当初条件分岐をつけないつもりだったが、followメソッドと全く同じ条件を追記し、return;とabort関数も併せて記述。
unfollowメソッドに条件分岐を加えたことで、ユーザー退会処理の際に条件に一致しないためunfollowメソッドを呼び出しできなくなったので、別途ユーザー退会のメソッドの中に、
退会時にフォローデータが削除されるロジックをリレーションメソッドを使って実装。
マイグレーションファイルの記述を直接カラムの生成に関係ない記述を一行を空けてまとめて記述した。
ユーザーを更新するルーティングを
ルートパラメーターでユーザーidを受取れるURLに修正。
ルートパラメーターを送るためにedit.blade.phpファイルのroute関数の第二引数にユーザーidを指定。そして不要になったinputタグを削除。
UserController内のupdateUserメソッドでrouteから渡されるユーザーidを変数で受け取れるように二つ目の引数の変数を実装。

Osk0830 commented on app/User.php:
```diff
@@ -52,4 +55,63 @@ public static function boot()
+     * @return view
+     */
+    public function followCheck ($id) {
+        $user = Auth::user();
+        return $user->followings()->where('followed_id', $id)->exists();
```
- フォローチェックは常にログインしているユーザーを基準に行われるものなので、そのインスタンスのメソッドどして実行されるかと思います。なので、こちらは再度インスタンスを呼び出す必要はなく、`$this->`でOKだと思われますがいかがでしょうか？

Osk0830 commented on resources/views/users/detail.blade.php:
```diff
@@ -8,10 +8,24 @@
+                            <div class="mt-3">
+                                <a href="{{ route('users.edit', $user->id) }}" class="btn btn-primary btn-block">ユーザ情報の編集</a>
+                            </div>
+                    @elseif ($user->followCheck($user->id))
```
- こちら、前回は見逃してしまいましたが、フォローチェックの基準となるユーザーは、この画面に詳細が表示されている `$user`ではなく、常にログインしているユーザーではないでしょうか？

Osk0830 commented on app/User.php:
```diff
@@ -52,4 +55,63 @@ public static function boot()
+     * @return view
+     */
+    public function unfollow($id) {
+        if (Auth::id() !== $id && !$this->followCheck($id)) {
```
- フォロー解除をしようとすると、スクショのように404エラーになります。おそらくここら辺の条件がおかしそうですね。フォローチェックの結果がどのような場合にアンフォローできるべきか再検討してみてくださいね。

[![Image from Gyazo](https://i.gyazo.com/c38c7f6fab2083744b487ed5ba71da85.gif)](https://gyazo.com/c38c7f6fab2083744b487ed5ba71da85)

Osk0830 commented on app/User.php:
```diff
@@ -52,4 +55,63 @@ public static function boot()
+     */
+    public function follow($id) {
+        DB::BeginTransaction();
+        if (Auth::id() !== $id && !$this->followCheck($id)) {
```
- パラメーターとして渡された値は文字列になります。`!==`を用いる場合は、型の問題にも一定程度配慮してあげましょう。

[![Image from Gyazo](https://i.gyazo.com/dcd7459f1706a3287e13bab5b9666857.png)](https://gyazo.com/dcd7459f1706a3287e13bab5b9666857)

[![Image from Gyazo](https://i.gyazo.com/a0cb8c74e8d340948b2dd9ac6c4eeee1.png)](https://gyazo.com/a0cb8c74e8d340948b2dd9ac6c4eeee1)

Osk0830 reviewed (changes requested):
お疲れ様です！
昨日まで諸事情によりドタバタしていた関係で遅くなってしまいました。

レビューしましたので、確認よろしくお願いします。

i-k-i-a-d committed:
followCheckメソッドを汎用的なメソッドにするためにを使って修正。これでログイン者でない人のフォロー確認も可能になった。
unfollowメソッドの条件式を修正。
新規投稿のphpDocを追記。
フォロー・フォロー解除ボタン表示の条件分岐をログイン者を基準としてフォロー状態をチェックして条件分岐するように修正。
マイグレーションファイルを修正し、created_atとupdated_atカラムを廃止。
follow_usersテーブル用のseederファイルの中身から作成日と更新日にダミーデータを入れる記述を削除。

Osk0830 reviewed (approved):
お疲れ様でした！
こちらしっかり修正できているのでマージしますね！
